Test for consistent use of icalmemory_new_buffer() et al and make the inner memory management functions configurable.

minichma opened the pull request:
This is a follow-up to libical/libical#542. It allows configuring the memory management functions (malloc, ...) used within `icalmemory` at runtime. The functionality is used to configure special versions of the functions during certain tests that ensure consistent use of memory allocation functions from icalmemory.

## `icalmemory`
The new functions `icalmemory_get_mem_alloc_funcs()` and `icalmemory_set_mem_alloc_funcs()` are introduced that allow getting and setting custom versions of `malloc()`, `realloc()` and  `free()`. The default versions are configured via `config.h.cmake`.

The changes are quite streight-forward and therefore not very risky.

## Tests
The regression tests are extended to use special test-versions of the mem-management functions that ensure consistent use of the memory management functions from `icalmemory()`. I.e. they ensure that:
* Memory allocated via `icalmemory` is not freed via stdlib `free()`, as otherwise an access violation would be raised.
* Memory allocated outside of `icalmemory` is not freed via `icalmemory`, as otherwise an assertion would be raised.

The test functions could also be used for verifying that the number of allocations matches the number of deallocations, but this would require some tweaking, as many tests load time zone data into the internal cache, which isn't freed at the end of each test as of today.

For now the test memory management funcs are only used in regression.c. It would certainly make sense to use them throughout all of the tests, but this would require some tweaking in the CMAKE build scripts. Any suggestion on how to best implement this in a fairly generic manner would be welcome.

minichma commented:
The build fails with errors I can't reproduce locally short term. Marking the PR as draft for now and will release it again as soon I could investigate and fix the issues.

winterz commented:
we could add a CMake option "MEMORY_CONSISTENCY'
and then in config.h.cmake wrap the ICALMEMORY_DEFAULT_FOO defines inside `if(MEMORY_CONSISTENCY)`

we should always check memory consistency with `make test`
right?

winterz commented:
cppcheck is saying:
`/home/allen/projects/libical/libical/src/test/test-malloc.c:110,error,memleakOnRealloc,Common realloc mistake: 'hdr' nulled but not freed upon failure`
is that a false positive?

winterz commented:
I can add the CMake stuff after this patch lands

winterz commented on src/test/test-malloc.c:
```diff
@@ -0,0 +1,179 @@
+#include "libical/ical.h"
+
+#include "regression.h"
+#include "test-malloc.h"
```
I like to include the functional unit's header above the others.
like so:
```
#include "test-malloc.h"
#include "regression.h"
```
I don't think `#include libical/ical.h` is necessary.

minichma commented:
> ppcheck is saying:
/home/allen/projects/libical/libical/src/test/test-malloc.c:110,error,memleakOnRealloc,Common realloc mistake: 'hdr' nulled but not freed upon failure
is that a false positive?

Yes, it is. I'll add inline suppressions.

minichma commented:
> we could add a CMake option "MEMORY_CONSISTENCY'
and then in config.h.cmake wrap the ICALMEMORY_DEFAULT_FOO defines inside if(MEMORY_CONSISTENCY)

The only downside would be, that the test wouldn't be active during 'regular' builds, not having the flag set. And the test-malloc is not designed to be used outside of test scenarios, so the built libs could only be used for testing if built with this flag.

> we should always check memory consistency with make test
right?

I think this would be good.

> I can add the CMake stuff after this patch lands

Would appreciate.

minichma commented on src/test/test-malloc.c:
```diff
@@ -0,0 +1,179 @@
+/*======================================================================
+FILE: regression-malloc.c
+
+(C) COPYRIGHT 2018, Markus Minichmayr
```
What is the preferred way of writing copyright headers? Do we have a notion of just writing 'The libical project' or similar and just adding the author as 'Initial Author'?

winterz commented:
still missing the include guard for `src/test/test-malloc.h`

regarding the CMake changes: 
 * -DMEMORY_CONSISTENCY can be a developer-only option.  similar to -DADDRESS_SANITIZER, -DTHREAD_SANITIZER, etc.  say in case someone wants to test it in a fuzzification or somesuch

* also should be regular memory consistency checking as part of the normal regression test suite.

does that make sense?

minichma commented:
Sorry for the missing header guard. Added it and cleaned up the commit history.

Regarding the CMAKE changes - sounds good. We would set these flags during the validation builds, right? I think, this would be important to enforce proper use of icalmemory functions.
